### PR TITLE
Endpoint for vault singleton creation

### PR DESCRIPTION
**Pull request by geoffwalmsley**

### CODE REVIEW

**Review by Quexington — DISMISSED** [review]

File: chia/rpc/wallet_rpc_api.py [hunk 1]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
...
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> EndpointResult:
+        """
+        Create a new vault
+        """
+        assert self.service.wallet_state_manager
+        wallet = self.service.wallet_state_manager.main_wallet
+
+        secp_pk = bytes.fromhex(str(request.get("secp_pk")))
+        entropy = bytes.fromhex(str(request.get("entropy")))
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4525]: Can we call this "hidden_puzzle_hash" now?

File: chia/rpc/wallet_rpc_api.py [hunk 2]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
         return {
             "transactions": [tx.to_json_dict_convenience(self.service.config) for tx in txs],
         }
+
+    ##########################################################################################
+    # VAULT
+    ##########################################################################################
+    @tx_endpoint(push=True)
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4511]: Let's start making these False by default, the ones with push=True currently are just for backwards compatbility

File: chia/rpc/wallet_rpc_api.py [hunk 3]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
...
+            additions=full_spend.additions(),
+            removals=full_spend.removals(),
+            wallet_id=wallet.id(),
+            sent_to=[],
+            trade_id=None,
+            type=uint32(TransactionType.INCOMING_TX.value),
+            name=full_spend.name(),
+            memos=[],
+            valid_times=ConditionValidTimes(),
+        )
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4582]: I'd like to keep the RPCs as pass-thru as possible so maybe this should all go into WSM

File: chia/rpc/wallet_rpc_api.py [hunk 4]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
...
+        # Get xch coin
+        amount = uint64(1)
+        coins = await wallet.select_coins(uint64(amount + fee), tx_config.coin_selection_config)
+
+        # Create singleton launcher
+        origin = coins.copy().pop()
+        launcher_coin = Coin(origin.name(), SINGLETON_LAUNCHER_PUZZLE_HASH, amount)
+
+        vault_full_puzzlehash = create_singleton_puzzle_hash(vault_puzzlehash, launcher_coin.name())
+        announcement_message = Program.to([vault_puzzlehash, amount, bytes(0x80)]).get_tree_hash()
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4542]: We should hint the secp_pk and hidden_puzzle_hash here so the information is not entirely lost to the chain

File: chia/rpc/wallet_rpc_api.py (hunk 1 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py:4525]: Actually, on further thought, let's take this out and come up with something canonical to use for the entropy of the hidden puzzle hash:
```
def calculate_hp_w_index_for_vault(index, hidden_puzzle = DEFAULT_HIDDEN_PUZZLE):
  return Program.to([6, (index, hidden_puzzle)])
```

Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py:4525]: I think we may need to have some way to make the entropy secret but we can add that later.

File: chia/rpc/wallet_rpc_api.py [hunk 5]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
...
+            fee,
+            coins,
+            None,
+            origin_id=origin.name(),
+            extra_conditions=(
+                AssertCoinAnnouncement(asserted_id=launcher_coin.name(), asserted_msg=announcement_message),
+            ),
+        )
+
+        genesis_launcher_solution = Program.to([vault_puzzlehash, amount, bytes(0x80)])
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4557]: Use the calculation you already did above

File: chia/rpc/wallet_rpc_api.py [hunk 6]
```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
...
+        launcher_cs = CoinSpend(launcher_coin, SINGLETON_LAUNCHER_PUZZLE, genesis_launcher_solution)
+        launcher_sb = SpendBundle([launcher_cs], AugSchemeMPL.aggregate([]))
+        assert tx_record.spend_bundle is not None
+        full_spend = SpendBundle.aggregate([tx_record.spend_bundle, launcher_sb])
+
+        vault_record = TransactionRecord(
+            confirmed_at_height=uint32(0),
+            created_at_time=uint64(int(time.time())),
+            amount=uint64(amount),
+            to_puzzle_hash=vault_full_puzzlehash,
```
Comment (Quexington) [inline chia/rpc/wallet_rpc_api.py:4568]: Use the inner puzzle hash you sent to here

File: tests/wallet/vault/test_vault_wallet.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    timelock = uint64(1000)
+    entropy = int_to_bytes(101)
+
+    res = await client.vault_create(SECP_PK, entropy, bls_pk, timelock, tx_config=DEFAULT_TX_CONFIG, fee=uint64(10))
+    vault_tx = res[0]
+    assert vault_tx
+
+    eve_coin = [item for item in vault_tx.additions if item not in vault_tx.removals and item.amount == 1][0]
+    assert eve_coin
```
Comment (Quexington) [inline tests/wallet/vault/test_vault_wallet.py:47]: Can we push the transaction and add a `.process_pending_states` call here to make sure the transaction goes through properly?

**Review by Quexington — CHANGES_REQUESTED** [review]

File: chia/wallet/wallet_state_manager.py [hunk 8]
```diff
@@ -2533,3 +2539,71 @@ async def sign_transaction(self, coin_spends: List[CoinSpend]) -> SpendBundle:
...
+
+        [tx_record] = await wallet.generate_signed_transaction(
+            amount,
+            SINGLETON_LAUNCHER_HASH,
+            tx_config,
+            fee,
+            coins,
+            None,
+            origin_id=origin.name(),
+            memos=[secp_pk, hidden_puzzle_hash],
```
Comment (Quexington) [inline chia/wallet/wallet_state_manager.py:2579]: One last nitpick, but these should go in the `genesis_launcher_solution` (where you currently have `None`) that way we don't have to fetch and run the spend before the launcher as well.

**Review by Quexington — APPROVED** [review]

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**